File: prototype/itk/Component.java
/**
 * itk – The Impressive Toolkit
 * 
 * Copyright © 2013  Mattias Andrée (maandree@member.fsf.org)
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package itk;

import java.awt.*;
import java.util.*;
import java.awt.image.BufferedImage;


/**
 * The root component class
 */
public class Component
{
    /**
     * The default background colour
     */
    private static final Color DEFAULT_BACKGROUND_COLOUR = new Color(211, 211, 211);
    
    
    
    /**
     * Constructor
     * 
     * @param  name  The name of the component
     */
    public Component(final String name)
    {
	this.name = name;
    }
    
    
    
    /**
     * The name of the component
     */
    public final String name;
    
    /**
     * The component's background colour
     */
    public Color backgroundColour = DEFAULT_BACKGROUND_COLOUR;
    
    /**
     * The component's advisory minimum size
     */
    public Dimension minimumSize = null;
    
    /**
     * The component's preferred size
     */
    public Dimension preferredSize = new Dimension(16, 16);
    
    /**
     * The component's size
     */
    public Dimension size = this.preferredSize;
    
    /**
     * The component's advisory maximum size
     */
    public Dimension maximumSize = null;
    
    /**
     * The components's parent
     */
    public Component parent = null;
    
    /**
     * The component's childred
     */
    public ArrayList<Component> children = new ArrayList<Component>();
    
    /**
     * Layout constraints
     */
    public Object constraints = null;
    
    /**
     * The component's layout manager
     */
    public LayoutManager layoutManager = null;
    
    /**
     * Buffer for double buffer
     */
    public volatile BufferedImage bufferImage = null;
    
    /**
     * Whether the component is double buffered
     */
    public boolean isDoubleBuffered = false;
    
    
    
    /**
     * Repaint the component and its childred
     * 
     * @param  g  The object with which to paint
     */
    public void paint(final Graphics2D g)
    {
	final int w = this.size.width, h = this.size.height;
	
	final boolean doubleBuffered = this.isDoubleBuffered;
	BufferedImage img = null;
	Graphics2D gg = g;
	if (doubleBuffered)
	{
	    img = bufferImage;
	    if ((img == null) || (img.getWidth() < w) || (img.getHeight() < h)) /* XXX new also of component shrunk considerably */
		this.bufferImage = img = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
	}
	
	this.paintComponent(gg);
	this.paintChildren(gg);
	
	if (doubleBuffered)
	    g.drawImage(img, 0, 0, null);
    }
    
    /**
     * Repaint the component
     * 
     * @param  g  The object with which to paint
     */
    protected void paintComponent(final Graphics2D g)
    {
	g.setColor(this.backgroundColour);
	g.fillRect(0, 0, this.size.width, this.size.height);
    }
    
    /**
     * Repaint the component's children
     * 
     * @param  g  The object with which to paint
     */
    protected void paintChildren(final Graphics2D g)
    {
	if (this.layoutManager != null)
	    this.layoutManager.prepare();
	try
	{
	    for (final Component child : this.children)
	    {
		final Rectangle rect = this.locateChild(child);
		if (rect != null)
		{
		    final int x = rect.x, width  = rect.width;
		    final int y = rect.y, height = rect.height;
		    child.paint((Graphics2D)(g.create(x, y, width, height)));
		}
	    }
	}
	finally
	{
	    if (this.layoutManager != null)
		this.layoutManager.done();
	}
    }
    
    
    /**
     * Synchronises the graphics
     */
    public void sync()
    {
	this.sync((Rectangle)null);
    }
    
    /**
     * Synchronises the graphics
     * 
     * @param  area  Area to synchronise, {@code null} for everything
     */
    public void sync(final Rectangle area)
    {
	final Graphics2D g;
	if (this.parent != null)
	{
	    if (this.backgroundColour.getAlpha() != 255)
	    {
		final Rectangle inParent = this.parent.locateChild(this);
		if (inParent != null)
		    if (area == null)
			this.parent.sync(inParent);
		    else
		    {
			final int x, y, w, h;
			x = area.x + inParent.x;
			y = area.y + inParent.y;
			w = area.width;
			h = area.height;
			this.parent.sync(new Rectangle(x, y, w, h));
		    }
	    }
	    else
		if ((g = this.parent.sync(this)) != null)
		{
		    if (area != null)
			g.clip(area);
		    this.paint(g);
		}
	}
    }
    
    /**
     * Synchronises the graphics on a child
     * 
     * @param   child  The child
     * @return         The object with which to paint
     */
    protected Graphics2D sync(final Component child)
    {
	if (this.parent == null)
	    return null;
	
	final Graphics2D g = this.parent.sync(this);
	if (g == null)
	    return null;
	
	final Rectangle rect = this.locateChild(child);
	if (rect == null)
	    return null;
	g.clip(rect);
	g.translate(-rect.x, -rect.y);
	return g;
    }
    
    
    /**
     * Locates the positions of the corners of a child
     * 
     * @param   child  The child
     * @return         The rectangle the child is confound in
     */
    public Rectangle locateChild(final Component child)
    {
	if (this.layoutManager != null)
	    return this.layoutManager.locate(child);
	
	int x = 0, y = 0;
	final int w = child.preferredSize.width, h = child.preferredSize.height;
	if ((child.constraints != null) && (child.constraints instanceof Point))
	{
	    x = ((Point)(child.constraints)).x;
	    y = ((Point)(child.constraints)).y;
	}
	
	child.size = child.preferredSize;
	return new Rectangle(x, y, w, h);
    }
    
    
    /**
     * {@inheritDoc}
     */
    @Override
    public String toString()
    {
	return this.name;
    }
    
}

File: src/component.c
/**
 * itk — The Impressive Toolkit
 * 
 * Copyright © 2013  Mattias Andrée (maandree@member.fsf.org)
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "component.h"
#include "layout_manager.h"
#include "graphics.h"
#include "itktypes.h"
#include "itkmacros.h"

#include <stdlib.h>

#define __this__  itk_component* this


/**
 * Locates the positions of the corners of a child
 * 
 * @param   child  The child
 * @return         The rectangle the child is confound in
 */
static rectangle_t locate_child(__this__, struct _itk_component* child)
{
  rectangle_t rc;
  
  if (this->layout_manager)
    return this->layout_manager->locate(this->layout_manager, child);
  
  rc.defined = true;
  rc.x = 0;
  rc.y = 0;
  rc.width = child->preferred_size.width;
  rc.height = child->preferred_size.height;
  if (child->constraints)
    {
      rc.x = ((position2_t*)(child->constraints))->x;
      rc.y = ((position2_t*)(child->constraints))->y;
    }
  
  child->size = child->preferred_size;
  return rc;
}


/**
 * Synchronises the graphics
 * 
 * @param  area  Area to synchronise, `NULL` for everything
 */
static void sync_area(__this__, rectangle_t* area)
{
  itk_graphics* g;
  rectangle_t rect;
  
  if (this->parent == NULL)
    return;
  
  if (this->background_colour.argb_colour.c.alpha != 255)
    {
      rect = this->parent->locate_child(this->parent, this);
      if (rect.defined && ((rect.width | rect.height) > 0))
	{
	  if ((area == NULL) || (area->defined == false) || ((area->width | area->height) < 0))
	    this->parent->sync_area(this->parent, &rect);
	  else
	    {
	      rect.x += area->x;
	      rect.y += area->y;
	      rect.width = area->width;
	      rect.height = area->height;
	      this->parent->sync_area(this->parent, &rect);
	    }
	}
    }
  else if ((g = this->parent->sync_child(this->parent, this)))
    {
      if ((area->defined) && ((area->width | area->height) >= 0))
	g->clip(g, *area);
      this->paint(this, g);
    }
}

/**
 * Synchronises the graphics on a child
 * 
 * @param   child  The child
 * @return         The object with which to paint
 */
static itk_graphics* sync_child(__this__, itk_component* child)
{
  if (this->parent == NULL)
    return NULL;
  
  itk_graphics* g = this->parent->sync_child(this->parent, this);
  if (g == NULL)
    return NULL;
  
  rectangle_t rect = this->locate_child(this, child);
  if ((rect.defined == false) && (rect.width | rect.height) < 0)
    return NULL;
  
  g->clip(g, rect);
  g->translate(g, new_position2(-(rect.x), -(rect.y)));
  return g;
}


/**
 * Synchronises the graphics
 */
static void sync(__this__)
{
  sync_area(this, NULL);
}


/**
 * Repaint the component and its childred
 * 
 * @param  g  The object with which to paint
 */
static void paint(__this__, itk_graphics* g)
{
  /* TODO implement buffer support */
  
  this->paint_component(this, g);
  this->paint_children(this, g);
}

/**
 * Repaint the component
 * 
 * @param  g  The object with which to paint
 */
static void paint_component(__this__, itk_graphics* g)
{
  g->set_colour(g, this->background_colour);
  g->fill_rectangle(g, new_rectangle(0, 0, this->size.width, this->size.height));
}

/**
 * Repaint the component's children
 * 
 * @param  g  The object with which to paint
 */
static void paint_children(__this__, itk_graphics* g)
{
  rectangle_t rect;
  itk_component* child;
  long i = 0, n = this->children_count;
  
  if (this->layout_manager)
    this->layout_manager->prepare(this->layout_manager);
  
  for (; i < n; i++)
    {
      child = *(this->children + i);
      rect = this->locate_child(this, child);
      if (rect.defined && (rect.width | rect.height) > 0)
	{
	  itk_graphics* child_g = g->create(g, rect);
	  child->paint(child, child_g);
	  free(child_g);
	}
    }
  
  if (this->layout_manager)
    this->layout_manager->done(this->layout_manager);
}


/**
 * Add a child component to the component
 * 
 * @param  child  The child
 */
static void add_child(__this__, itk_component* child)
{
  if (this->children_count == 0)
    this->children = malloc(sizeof(itk_component*));
  else if (this->children_count & -(this->children_count) == this->children_count)
    this->children = realloc(this->children, this->children_count * 2 * sizeof(itk_component*));
  *(this->children + this->children_count++) = child;
}

/**
 * Remove a child component from the component
 * 
 * @param  child  The child
 */
static void remove_child(__this__, itk_component* child)
{
  long i;
  for (i = 0;; i++)
    if (*(this->children + i) == child)
      {
	this->remove_child_by_index(this, i);
	break;
      }
}

/**
 * Remove a child component from the component
 * 
 * @param  child  The index of the child
 */
static void remove_child_by_index(__this__, long child)
{
  long i, n = this->children_count -= 1;
  for (i = child; i < n; i++)
    *(this->children + i) = *(this->children + i + 1);
  if (this->children_count == 0)
    {
      free(this->children);
      this->children = NULL;
    }
  else if (this->children_count & -(this->children_count) == this->children_count)
    this->children = realloc(this->children, this->children_count * sizeof(itk_component*));
}


/**
 * Destructor
 */
void free_component(__this__)
{
  free(this);
}

  
/**
 * Destructor that also frees the layout manager and children
 */
void free_everything_component(__this__)
{
  long i = 0, n = this->children_count;
  
  if (this->layout_manager)
    this->layout_manager->free(this->layout_manager);
  
  for (; i < n; i++)
    (*(this->children + i))->free_everything(*(this->children + i));
  
  this->free(this);
}


/**
 * Constructor
 * 
 * @param  name  The name of the component
 */
itk_component* itk_new_component(char* name)
{
  itk_component* rc = calloc(1, sizeof(itk_component));
  rc->name = name;
  rc->visible = 1;
  rc->background_colour = new_colour_by_opaque_value(0xD3D3D3);
  rc->preferred_size = new_size2(16, 16);
  rc->size = new_size2(16, 16);
  rc->maximum_size = new_size2(UNBOUNDED, UNBOUNDED);
  rc->locate_child = locate_child;
  rc->sync = sync;
  rc->sync_area = sync_area;
  rc->sync_child = sync_child;
  rc->paint = paint;
  rc->paint_component = paint_component;
  rc->paint_children = paint_children;
  rc->add_child = add_child;
  rc->remove_child = remove_child;
  rc->remove_child_by_index = remove_child_by_index;
  rc->free = free_component;
  rc->free_everything = free_everything_component;
  return rc;
}

